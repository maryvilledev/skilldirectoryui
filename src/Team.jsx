import React from 'react';
import 'bootstrap/dist/css/bootstrap.css';
import Button from 'react-bootstrap/lib/Button';
import Modal from 'react-modal';
import { Row, Col }  from 'react-bootstrap';
import axios from 'axios';
import Select from 'react-select';
import { ModalStyle } from './Styles'

import AddTeamMemberForm from './AddTeamMemberForm.jsx';
import SelectedItem from './SelectedItem.jsx';
import TeamMemberDisplay from './TeamMemberDisplay.jsx';
import DeleteModal from './DeleteModal.jsx';
import DeleteButton from './DeleteButton.jsx';

var api = (process.env.REACT_APP_API);

class Team extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      teamMembers: [],
      addModalIsOpen: false,
      deleteModalIsOpen: false,
      selectedTeamMember: {
        id: "",
        name: "",
        title: "",
      }
    };

    this.fetchTeamMembers = this.fetchTeamMembers.bind(this);
    this.openAddModal = this.openAddModal.bind(this);
    this.closeAddModal = this.closeAddModal.bind(this);
    this.openDeleteModal = this.openDeleteModal.bind(this);
    this.closeDeleteModal = this.closeDeleteModal.bind(this);
    this.onSelectChange = this.onSelectChange.bind(this);
    this.addTeamMember = this.addTeamMember.bind(this);
    this.shouldDelete = this.shouldDelete.bind(this);
    this.deleteTeamMember = this.deleteTeamMember.bind(this);
  }

  openAddModal() {
    this.setState({ addModalIsOpen: true });
  }

  closeAddModal() {
    this.setState({ addModalIsOpen: false });
  }

  openDeleteModal() {
    this.setState({ deleteModalIsOpen: true });
  }

  closeDeleteModal() {
    this.setState({ deleteModalIsOpen: false });
  }

  fetchTeamMembers() {
    axios.get(`${api}/teammembers/`)
      .then((result) => {
        const teamMembers = result.data.slice();
        this.setState({
          teamMembers: teamMembers
        });
      })
      .catch((err) => {
        console.log(`Error fetching team members: ${err}`);
      });
  }

  onSelectChange(obj) {
    axios.get(`${api}/teammembers/${obj.id}`)
      .then((result) => {
        const teamMemberData = result.data;
        this.setState({
          selectedTeamMember: {
            id: teamMemberData.id,
            name: teamMemberData.name,
            title: teamMemberData.title,
          }
        });
      }).catch((err) => {
        console.log(`Error: ${err}`);
      });
  }

  addTeamMember(name, title) {
    axios.post(`${api}/teammembers/`, {
      name: name,
      title: title
    }).then((response) => {
        this.closeAddModal();
        this.fetchTeamMembers();
      }).catch(err => {
        console.log(`Error: ${err}`);
      });
  }

  shouldDelete(response) {
    if (response) {
      this.deleteTeamMember();
    }
    this.closeDeleteModal();
  }

  deleteTeamMember() {
    axios.delete(`${api}/teammembers/${this.state.selectedTeamMember.id}`)
      .then((result) => {
        this.setState({
          selectedTeamMember: {
            id: "",
            name: "",
            title: "",
          }
        });
        // Refetch the list of team members
        this.fetchTeamMembers();
      })
      .catch((err) => {
        console.log(`Error: ${err}`);
      });
  }

  componentDidMount() {
    this.fetchTeamMembers();
  }

  render() {
    let selectedItem = null;
    if (this.state.selectedTeamMember.id) {
      selectedItem = (
        <SelectedItem
          typeName="TeamMember"
          deleteCallback={this.openDeleteModal}
        >
          <TeamMemberDisplay
            selected={this.state.selectedTeamMember}
          />
        </SelectedItem>
      );
    }
    return (
      <div>
        <Row>
          <Col xs={4} md={4}>
            <Select
              name="teamMembers"
              labelKey="name"
              value={this.state.selectedTeamMember.name}
              options={this.state.teamMembers}
              onChange={this.onSelectChange} />
          </Col>
          <Button
            name="addTeamMember"
            bsStyle="primary"
            onClick={this.openAddModal}
            children="Add Team Member" />

          <Modal
            isOpen={this.state.addModalIsOpen}
            onRequestClose={this.closeAddModal}
            contentLabel="AddTeamMemberModal"
            style={ModalStyle} >
            <AddTeamMemberForm
              onSubmit={this.addTeamMember}
              closeModal={this.closeAddModal}/>
          </Modal>
        </Row>
        {selectedItem}
        <Modal
          isOpen={this.state.deleteModalIsOpen}
          onRequestClose={this.closeDeleteModal}
          contentLabel="DeleteTeamMemberModal"
          style={ModalStyle} >
          <DeleteModal
            shouldDelete={this.shouldDelete}
          />
        </Modal>
      </div>
    );
  }
}

export default Team;
