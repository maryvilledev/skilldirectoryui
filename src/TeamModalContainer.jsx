import React, { PropTypes } from 'react';
import Modal from 'react-modal';

import AddTeamMemberForm from './AddTeamMemberForm.jsx';
import DeleteModal from './DeleteModal.jsx';
import { ModalStyle } from './Styles.jsx';

// Object to map displayedModalType to a component
const teamModalTypes = {
  AddTeamMember: AddTeamMemberForm,
  DeleteTeamMember: DeleteModal,
};

const TeamModalContainer = ({ closeModalCallback, displayedModalType, formProps, isModalDisplayed}) => {
  const ModalToDisplay = teamModalTypes[displayedModalType];
  const modalComponent = (ModalToDisplay ?
    <ModalToDisplay {...formProps} /> : null);
  return (
    <Modal
      isOpen={isModalDisplayed}
      onRequestClose={closeModalCallback}
      contentLabel="TeamModalContainer"
      style={ModalStyle}
    >
      {modalComponent}
    </Modal>
  );
};

TeamModalContainer.propTypes = {
  closeModalCallback: PropTypes.func,
  displayedModalType: PropTypes.string,
  formProps: PropTypes.object,
  isModalDisplayed: PropTypes.bool,
};

TeamModalContainer.defaultProps = {
  closeModalCallback: () => {},
  displayedModalType: '',
  formProps: {},
  isModalDisplayed: false,
};

export default TeamModalContainer;
