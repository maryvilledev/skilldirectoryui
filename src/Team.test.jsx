import React from 'react';
import ReactDOM from 'react-dom';
import cookie from 'react-cookie';
import { shallow, mount } from 'enzyme';

import Team from './Team.jsx';
import TeamModalContainer from './TeamModalContainer.jsx';

describe('<Team />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Team />, div);
  });

  it('matches the stored snapshot', () => {
    const wrapper = shallow(<Team />);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  describe('when the user is not logged in', () => {
    beforeEach(() => {
      cookie.save('isLoggedIn', false);
    });
    it('should not display an Add Team Member button', () => {
      const wrapper = mount(<Team />);
      const addTeamMemberButton = wrapper.find('[name="addTeamMember"]');
      expect(addTeamMemberButton).toHaveLength(0);
    });
    it('does not appear when no team member is selected', () => {
      const wrapper = mount(<Team />);
      const deleteButton = wrapper.find('[name="DeleteTeamMember"]');
      expect(deleteButton).toHaveLength(0);
    });
  });

  describe('when the user is logged in', () => {
    beforeEach(() => {
      cookie.save('isLoggedIn', true);
    });

    afterEach(() => {
      cookie.remove('isLoggedIn');
    });

    describe('the Add Team Member button', () => {
      it('should be displayed', () => {
        const wrapper = mount(<Team />);
        const addTeamMemberButton = wrapper.find('[name="addTeamMember"]');
        expect(addTeamMemberButton).toHaveLength(1);
      });
      it('should open an AddTeamMemberModal when clicked', () => {
        const wrapper = mount(<Team />);
        const teamModalContainer = wrapper.find(TeamModalContainer);
        const addTeamMemberButton = wrapper.find('[name="addTeamMember"]');
        expect(teamModalContainer.prop('isModalDisplayed')).toBe(false);
        expect(teamModalContainer.prop('displayedModalType')).toBe('');
        addTeamMemberButton.simulate('click');
        expect(teamModalContainer.prop('isModalDisplayed')).toBe(true);
        expect(teamModalContainer.prop('displayedModalType')).toBe('AddTeamMember');
      });
    });
    describe('the delete button', () => {
      it('appears when a team member is selected', () => {
        const wrapper = mount(<Team />);
        wrapper.setState({ selectedTeamMember: { id: '1' } });
        const deleteButton = wrapper.find('[name="DeleteTeamMember"]');
        expect(deleteButton).toHaveLength(1);
      });
      it('opens a DeleteModal when clicked', () => {
        const wrapper = mount(<Team />);
        wrapper.setState({ selectedTeamMember: { id: '1' } });
        const teamModalContainer = wrapper.find(TeamModalContainer);
        const deleteButton = wrapper.find('[name="DeleteTeamMember"]');
        expect(teamModalContainer.prop('isModalDisplayed')).toBe(false);
        expect(teamModalContainer.prop('displayedModalType')).toBe('');
        deleteButton.simulate('click');
        expect(teamModalContainer.prop('isModalDisplayed')).toBe(true);
        expect(teamModalContainer.prop('displayedModalType')).toBe('DeleteTeamMember');
      });
    });
  });
});
