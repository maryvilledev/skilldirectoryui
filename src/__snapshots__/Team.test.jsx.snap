exports[`<Team /> matches the stored snapshot 1`] = `
<div>
  <Row
    bsClass="row"
    componentClass="div">
    <Col
      bsClass="col"
      componentClass="div"
      md={4}
      xs={4}>
      <Select
        addLabelText="Add \"{label}\"?"
        arrowRenderer={[Function]}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        clearAllText="Clear all"
        clearValueText="Clear value"
        clearable={true}
        delimiter=","
        disabled={false}
        escapeClearsValue={true}
        filterOptions={[Function]}
        ignoreAccents={true}
        ignoreCase={true}
        inputProps={Object {}}
        isLoading={false}
        joinValues={false}
        labelKey="name"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function]}
        multi={false}
        name="teamMembers"
        noResultsText="No results found"
        onBlurResetsInput={true}
        onChange={[Function]}
        onCloseResetsInput={true}
        openAfterFocus={false}
        optionComponent={[Function]}
        options={Array []}
        pageSize={5}
        placeholder="Select..."
        required={false}
        scrollMenuIntoView={true}
        searchable={true}
        simpleValue={false}
        tabSelectsValue={true}
        value=""
        valueComponent={[Function]}
        valueKey="value" />
    </Col>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      name="addTeamMember"
      onClick={[Function]}>
      Add Team Member
    </Button>
    <Modal
      ariaHideApp={true}
      closeTimeoutMS={0}
      contentLabel="AddTeamMemberModal"
      isOpen={false}
      onRequestClose={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      shouldCloseOnOverlayClick={true}
      style={
        Object {
          "content": Object {},
          "overlay": Object {
            "bottom": "25%",
            "left": "25%",
            "position": "fixed",
            "right": "25%",
            "top": "15%",
          },
        }
      }>
      <AddTeamMemberForm
        closeModal={[Function]}
        onSubmit={[Function]} />
    </Modal>
  </Row>
  <Modal
    ariaHideApp={true}
    closeTimeoutMS={0}
    contentLabel="DeleteTeamMemberModal"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnOverlayClick={true}
    style={
      Object {
        "content": Object {},
        "overlay": Object {
          "bottom": "25%",
          "left": "25%",
          "position": "fixed",
          "right": "25%",
          "top": "15%",
        },
      }
    }>
    <DeleteModal
      shouldDelete={[Function]} />
  </Modal>
</div>
`;
